apiclient/census.go

```diff
@@ -112,3 +133,113 @@ func (c *HTTPclient) CensusGenProof(censusID, voterKey types.HexBytes) (*CensusP
 	}
 	return &cp, nil
 }
+
+// CensusGenProofZk function generates the census proof of a election based on
+// ZkSnarks. It uses the current apiclient circuit config to instance the
+// circuit and generates the proof for the censusRoot, electionId and voter
+// babyjubjub private key provided.
+func (c *HTTPclient) CensusGenProofZk(censusRoot, electionId types.HexBytes) (*CensusProofZk, error) {
+	// Perform a request to get the type of the current census and check if it
+	// is a zk census.
+	resp, code, err := c.Request("GET", nil, "censuses", censusRoot.String(), "type")
+	if err != nil {
+		return nil, err
+	}
+	if code != 200 {
+		return nil, fmt.Errorf("%s: %d (%s)", errCodeNot200, code, resp)
+	}
+	censusData := &api.Census{}
+	err = json.Unmarshal(resp, censusData)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal response: %w", err)
+	}
+
+	if censusData.Type != api.CensusTypeZKWeighted {
+		return nil, fmt.Errorf("current census is not a zk census")
+	}
+
+	// Get merkle proof associated to the voter key provided, that will contains
+	// the leaf siblings and value (weight)
+	resp, code, err = c.Request("GET", nil, "censuses", censusRoot.String(), "proof", c.zkAddr.String())
+	if err != nil {
+		return nil, err
+	}
+	if code != 200 {
+		return nil, fmt.Errorf("%s: %d (%s)", errCodeNot200, code, resp)
+	}
+	censusData = &api.Census{}
+	err = json.Unmarshal(resp, censusData)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal response: %w", err)
+	}
+
+	log.Debugw("zk census data received, starting to generate the proof inputs...", map[string]interface{}{
+		"censusRoot": censusRoot.String(),
+		"electionId": electionId.String(),
+		"zkAddress":  c.zkAddr.String(),
+	})
+
+	// get nullifier
+	nullifier, err := c.zkAddr.Nullifier(electionId)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode census root
+	strCensusRoot := arbo.BytesToBigInt(censusRoot).String()
+	// Get vote weight
+	weight := new(big.Int).SetInt64(1)
+	if censusData.Weight != nil {
+		weight = censusData.Weight.MathBigInt()
+	}
+
+	// Calculate and encode vote hash -> sha256(voteWeight)
+	encElectionId := zk.BytesToArboStr(electionId)
+	voteHash := zk.BytesToArboStr(weight.Bytes())
+
+	// Create the inputs and encode them into a JSON
+	rawInputs := map[string]interface{}{
```

Comment:
I think this struct should be declared somewhere else. Probably in the crypto/zk package.

Reply:
here @lucasmenendez